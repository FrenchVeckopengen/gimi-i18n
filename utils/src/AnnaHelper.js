//@flow
var symbolsThatShouldNotBeReplaced = ['%1$s', '%2$s', '%3$s', '%4$s', '%1$d', '%2$d', '%3$d', '%4$d', '%5$d', '%6$d', '%7$d', '%8$d', 'PLZ_TRANSLATE', "'"]

let toHash = (text: string) => {
  symbolsThatShouldNotBeReplaced.forEach((symbol, index) => {
    var splits = text.split(symbol)
    text = splits.join(`<span class="notranslate">${symbol}</span>`)
  })

  //support server strings
  text = text.replace(/{/g,'<span class="notranslate">{');
  text = text.replace(/}/g, '}</span>')

  return text
}

let fromHash = (text: string) => {
  text = text.replace(/<span class="notranslate">/g,'');
  text = text.replace(/<\/span>/g, '')
  return text
}


let translationHelpTemplate = 'PLZ_TRANSLATE'
let translationTravisHelper = 'ANNA'
module.exports = {toHash, fromHash, translationHelpTemplate, translationTravisHelper}
