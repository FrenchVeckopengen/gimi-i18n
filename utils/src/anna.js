/* eslint no-console: ["error", { allow: ["warn", "error", "log"] }] */

// @martin, alright, i would first do this as a seperate project called anna becuase we dont want to screw up sara and this might not work
var fs = require('fs')
var translate = require('@google-cloud/translate')({
  projectId: 'gimi-969a7',
  keyFilename: './utils/gkey/Gimi-b4f63676ca99.json'
})

var AnnaHelper = require('./AnnaHelper')
let templateDir = ['./text_strings/client', './text_strings/server', './text_strings/templates']

let RunAnna = (filePath):* => {
  let getPath = (file) => `${filePath}/${file}`
  var {toHash, fromHash, translationHelpTemplate} = AnnaHelper
  let translateTextStringForFile = (file, textId) => {
    if (file === 'default.json') { return Promise.resolve() }
    if (file.indexOf('.json') === -1) { return Promise.resolve() }

    if (file === 'sv.json') {
      return Promise.resolve()
    }

    var path = getPath(file)
    var TextStrings = fs.readFileSync(path, {encoding: 'utf8'})
    TextStrings = JSON.parse(TextStrings)
    var stringToTranslate = TextStrings[textId]

    if (!stringToTranslate) { return Promise.reject(new Error(`Cant find textid: ${textId} in file: ${path}`)) }
    var lang = file.replace('TextStrings_', '').replace('.json', '')
    if (lang === 'nb') {
      lang = 'no'
    }

    // hash %1$d to num evals to work with google translate
    stringToTranslate = toHash(stringToTranslate)

    return translate.translate(stringToTranslate, lang, (err, translation) => {
      if (err) {
        return console.log(`Error on textId ${textId}. message: ${err}. file: ${file}`)
      }
      var translatedText = fromHash(translation)
      console.log(`Translated text: '${stringToTranslate}' to: '${translatedText}' in ${filePath}/${file}`)
      var re = /\*\*\*(.*?)\*\*\*/g
      if (re.test(translatedText)) {
        TextStrings[textId] = translatedText
      } else {
        TextStrings[textId] = translationHelpTemplate + translatedText
      }

      TextStrings = JSON.stringify(TextStrings, undefined, 2)
      fs.unlinkSync(path)
      return fs.writeFileSync(path, TextStrings, {encoding: 'utf8'})
    })
  }

  let textIdToTranslate = process.argv[2]

  if (!textIdToTranslate) { console.log('use: npm run anna -- <text_id>') }

  if (textIdToTranslate) {
    Promise.all(fs.readdirSync(filePath).map((file) => translateTextStringForFile(file, textIdToTranslate)))
      .then(() => console.log('saved Successfully :)'))
      .catch((err) => console.error(err))
  }
}
templateDir.forEach((filePath) => {
  RunAnna(filePath)
})
