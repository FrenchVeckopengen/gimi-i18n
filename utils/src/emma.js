
const {execSync} = require('child_process')
var fs = require('fs')
var {TRAVIS} = process.env

let templateDir = ['./text_strings/client', './text_strings/server', './text_strings/templates', './text_strings/gimi-web', './text_strings/share-image-generator']
let files = ['sv.json', 'en.json']

var LADIES = {
  anna: 'ANNA',
  emma: 'EMMA',
  poli: 'POLI',
  krisi: 'KRISI'
}

let getTranslateInfoFrom = (path) => {
  var TextStringsStr = fs.readFileSync(path, {encoding: 'utf8'})
  var TextStrings = JSON.parse(TextStringsStr)
  var keys = Object.keys(TextStrings)
  let textIds = []

  keys.forEach(textid => {
    var text = TextStrings[textid]
    if (!text) text = ''

    Object.keys(LADIES).forEach((lady) => {
      if (TextStrings[textid].includes(LADIES[lady])) textIds.push({textid, lady})
    })
  })

  return {textIds, path}
}

let removeANNAStringFromFile = (path) => {
  var json = fs.readFileSync(path, {encoding: 'utf8'})
  Object.keys(LADIES).forEach((lady) => {
    let LADY = LADIES[lady]
    json = json.replace(new RegExp(`${LADY} `, 'g'), '')
    json = json.replace(new RegExp(` ${LADY}`, 'g'), '')
    json = json.replace(new RegExp(`${LADY}`, 'g'), '')
  })

  fs.unlinkSync(path)
  fs.writeFileSync(path, json, {encoding: 'utf8'})
}

var info = templateDir.reduce((a, b) => a.concat(files.map((file) => getTranslateInfoFrom(`${b}/${file}`))), [])
info = info.filter((x) => x.textIds.length > 0)

info.forEach(x => removeANNAStringFromFile(x.path))

info.forEach((x) => {
  var {textIds} = x
  textIds.map(({textid, lady}) => {
    let cmd = `npm run ${lady}-d ${textid}`
    // eslint-disable-next-line
    console.log(cmd)
    execSync(cmd)
  })
})

if (info.length === 0) {
  // eslint-disable-next-line
  console.log('Found no textstrings to translate')
  process.exit()
}
// eslint-disable-next-line
if (!process.argv.join().includes('--push-changes')) process.exit()

// eslint-disable-next-line
execSync('git remote add origin-with-push-access https://${GH_TOKEN}@github.com/gimi-org/gimi-i18n.git')
execSync('git add --all')
execSync(`git commit -m "Google Translated ${info[0].textIds.join(' ')}"`)
if (TRAVIS) execSync('git push --no-verify --set-upstream origin-with-push-access')
