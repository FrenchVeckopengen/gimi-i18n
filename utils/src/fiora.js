var Slack = require('node-slack')
var slack = new Slack('https://hooks.slack.com/services/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')

export var languageCodes = ['da', 'fi', 'is', 'sv', 'nb', 'en', 'fr', 'nl', 'be', 'it', 'es', 'de', 'et', 'is']

var textStringSets = {}
languageCodes.forEach((code) => (textStringSets[code] = require(`../../text_strings/client/${code}.json`)))

export let checkStringLenght = (firstLang: Object, secondLang: Object, firstLangName: string, secondLangName: string):Array<Object> => {
  var keys = Object.keys(firstLang)

  var longTextWarning = []
  var longTextSlackData = []
  keys.forEach(key => {
    if (firstLangName === 'en' && secondLangName !== 'en') {
      if (secondLang[key].includes('PLZ_TRANSLATE')) secondLang[key] = secondLang[key].replace('PLZ_TRANSLATE', '')
      var differencePerc = (firstLang[key].length - secondLang[key].length) / 100

      if (Math.abs(differencePerc) >= 0.20) {
        var actIncr = Math.abs(differencePerc)
        longTextWarning.push(`Lang: ${firstLangName}, Key: ${key} is 20% longer than: ${secondLangName} -> ${actIncr}`)
        longTextSlackData.push({lang: firstLangName, title: key, secLang: secondLangName, value: differencePerc, short: true})
      }
    }
    return true
  })
  // eslint-disable-next-line
  if (longTextWarning.length >0) {console.warn(longTextWarning)}
  return longTextSlackData
}

let runFiora = ():* => {
  var stringLengthData = languageCodes.forEach(lang2 => checkStringLenght(textStringSets.en, textStringSets[lang2], 'en', lang2))

  var attachmentPayload = [{
    'fallback': 'Required text summary of the attachment that is shown by clients that understand attachments but choose not to show them.',
    'text': 'Optional text that should appear within the attachment',
    'pretext': 'Optional text that should appear above the formatted data',
    'color': '#36a64f', // Can either be one of 'good', 'warning', 'danger', or any hex color code
    // Fields are displayed in a table on the message
    'fields': [
      {
        'title': 'Required Field Title', // The title may not contain markup and will be escaped for you
        'value': 'Text value of the field. May contain standard message markup and must be escaped as normal. May be multi-line.',
        'short': false // Optional flag indicating whether the `value` is short enough to be displayed side-by-side with other values
      }
    ]
  }, {
    'fallback': 'String length information',
    'color': '#36a64f',
    'fields': stringLengthData
  }]
  slack.send({
    text: 'Greetings Here is todays info!',
    channel: '#translation_status',
    username: 'TranslationStatus',
    icon_emoji: 'bread',
    attachments: attachmentPayload,
    unfurl_links: true,
    link_names: 1
  })
}
runFiora()
