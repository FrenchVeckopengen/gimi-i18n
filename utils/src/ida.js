
/* eslint no-console: ["error", { allow: ["warn", "error", "log"] }] */
var fs = require('fs')

let RunIda = (): * => {
  var files = ['config_AUD.json', 'config_CAD.json', 'config_DEF.json', 'config_DKK.json', 'config_EUR.json', 'config_GBP.json', 'config_INR.json', 'config_ISK.json', 'config_NOK.json',
    'config_NZD.json', 'config_SEK.json', 'config_THB.json', 'config_USD.json', 'config_IDR.json', 'config_CHF.json']
  files.forEach((file) => {
    syncConfigKeys(file)
  })
}

let syncConfigKeys = (file: string) => {
  var sekConfig = require('../../config/config_SEK.json')
  var config = require('../../config/' + file)

  // Delete Support
  Object.keys(config)
    .filter((key) => sekConfig[key] === undefined)
    .forEach((key) => {
      delete config[key]
      console.log(`Deleting key: '${key}' from ${file}`)
    })

  // Create
  Object.keys(sekConfig)
    .filter((key) => config[key] === undefined)
    .forEach((key) => {
      config[key] = sekConfig[key]
      console.log(`Creating key: '${key}' in ${file}`)
    })

  // Save changes
  var newConf = JSON.stringify(config, undefined, 2)
  fs.unlinkSync('./config/' + file)
  fs.writeFileSync('./config/' + file, newConf, {encoding: 'utf8'})
}

RunIda()
