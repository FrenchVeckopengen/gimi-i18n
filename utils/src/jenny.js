/* eslint flow-header/flow-header: 0 */
/* eslint no-console: ["error", { allow: ["warn", "error", "log"] }] */
var keysToIgnore = [
  'child_about_data_tracking_title',
  'child_about_data_tracking_content',
  'parent_about_data_tracking_title',
  'parent_about_data_tracking_content',
  'child_advice_title',
  'child_advice_content',
  'parent_advice_title',
  'parent_advice_content',
  'data_knowledge_item_text',
  'knowledge_child_item_content',
  'knowledge_child_item_header',
  'data_tracking_assignment_threshold',
  '_permissions_denied',
  'vp_help_page_title',
  'vp_help_page_content',
  'transactions_help_page_title',
  'transactions_help_page_content',
  'task_editor_title_input_hint_example',
  'task_status',
  'category_name_1',
  'category_name_2',
  'category_name_3',
  'category_name_4',
  'category_name_5',
  'category_name_6',
  'category_name_7',
  'category_name_8',
  'category_name_9',
  'category_name_10',
  'category_name_11',
  'category_name_12',
  'category_name_13',
  'category_name_14',
  'category_name_15',
  'category_name_16',
  'category_name_17',
  'category_name_18',
  'category_name_19',
  'category_name_20',
  'category_name_21',
  'category_name_22',
  'category_name_23',
  'category_name_24',
  'category_name_25',
  'category_name_26',
  'category_name_27',
  'category_name_28',
  'category_name_29',
  'category_name_30',
  'category_name_31',
  'category_name_32',
  'category_name_33',
  'category_name_34',
  'category_name_35',
  'category_name_36',
  'category_name_37',
  'category_name_38',
  'category_name_39',
  'category_name_40',
  'category_name_41',
  'category_name_42',
  'category_name_43',
  'category_name_44',
  'category_name_45',
  'category_name_46',
  'category_name_47',
  'category_name_48'
]

var keysToDelete = []
var fs = require('fs')
var rootDir = '..'
var dirsToCheck = ['components', 'libs', 'hocs', 'i18n', 'config', 'reducers']
let textStringsSvFilePath = './text_strings/client/sv.json'
var TextStrings = fs.readFileSync(textStringsSvFilePath, {encoding: 'utf8'})
TextStrings = JSON.parse(TextStrings)
var foundKeys = 0
var ignoredKeys = 0

let checkFile = (file, key) => {
  var fileContents = fs.readFileSync(file, {encoding: 'utf8'})

  var isOk = false

  if (fileContents.indexOf(key) !== -1) isOk = true

  if (key.indexOf('_parent') !== -1) {
    if (fileContents.indexOf(key.split('_parent')[0]) !== -1) isOk = true
  }

  if (key.indexOf('_child') !== -1) {
    if (fileContents.indexOf(key.split('_child')[0]) !== -1) isOk = true
  }

  return isOk
}
let checkIfTextStringIsObsolete = (key) => {
  var isOk = false
  dirsToCheck.forEach((dirName) => {
    var dir = `${rootDir}/${dirName}/`
    var files = fs.readdirSync(dir)
    files.forEach((file) => {
      try {
        file = `${dir}/${file}`
        // console.log(`Reading file: "${file}"`)
        if (checkFile(file, key)) {
          isOk = true
        }
      } catch (e) {
        // console.log(`Failed reading file: "${file}"`)
        // console.log(e.message)
      }
    })
  })

  // text string is probably obsolete
  if (!isOk) {
    if (keysToIgnore.some(ignoredKey => key.indexOf(ignoredKey) !== -1)) { ignoredKeys++ } else {
      foundKeys++
      keysToDelete.push(key)
      console.log(`${key}\t\t\t\t\t\t${TextStrings[key].replace('\n', '')}`)
    }
  }
}

console.log('****** Begin Scan ********\n')

Object.keys(TextStrings).forEach(key => checkIfTextStringIsObsolete(key))

console.log('****** Scan Complete ********')
console.log(
  `
  Ignored text_strings: ${ignoredKeys}
  Found text_strings to delete: ${foundKeys}`)
if (process.argv.some(x => x === 'f')) {
  /** ****** Deleting TextStrings *******/
  console.log(`
      Removing ${foundKeys} text_strings from ${textStringsSvFilePath} ..
      `)
  keysToDelete.forEach((key) => {
    delete TextStrings[key]
  })

  TextStrings = JSON.stringify(TextStrings, undefined, 2)
  fs.unlinkSync(textStringsSvFilePath)
  fs.writeFileSync(textStringsSvFilePath, TextStrings, {encoding: 'utf8'})

  console.log(`
      Done
      `)
} else {
  console.log(`
  run with f to remove ${foundKeys} text_strings from ${textStringsSvFilePath}
  `)
}
