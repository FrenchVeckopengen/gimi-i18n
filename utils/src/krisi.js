/* eslint no-console: ["error", { allow: ["warn", "error", "log"] }] */

var fs = require('fs')
var translate = require('@google-cloud/translate')({
  projectId: 'gimi-969a7',
  keyFilename: './utils/gkey/Gimi-b4f63676ca99.json'
})

var AnnaHelper = require('./AnnaHelper')
let templateDir = ['./text_strings/client']

let RunPolina = (filePath):* => {
  let getPath = (languageFile) => `${filePath}/${languageCode}`
  var {toHash, fromHash, translationHelpTemplate} = AnnaHelper
  let translateLanguageFile = (languageFile) => {
    if (languageFile.indexOf('.json') === -1) { return void 0 }

    var path = getPath(languageFile)
    var TextStrings = fs.readFileSync(path, {encoding: 'utf8'})
    // TextStrings = JSON.parse(TextStrings)

    var textStringArray = TextStrings.split(':')
    for (var i = 0; i < TextStrings.length; i++) {
      console.log(textStringArray[0])
      var textId = textStringArray[0]
      var stringToTranslate = textStringArray[0]

      if (!stringToTranslate) {
        console.warn(`Cant find: ${languageCode} in file: ${path}`)
        return void 0
      }
      var lang = languageFile.replace('TextStrings_', '').replace('.json', '')
      // hash %1$d to num evals to work with google translate
      stringToTranslate = toHash(stringToTranslate)
      i++

      translate.translate(stringToTranslate, lang, (err, translation) => {
        if (err) {
          return console.log(`Error on textId ${textId}. message: ${err}. file: ${languageFile}`)
        }
        var translatedText = fromHash(translation)
        console.log(`Translated text: '${stringToTranslate}' to: '${translatedText}' in ${filePath}/${languageFile}`)
        var re = /PLZ_TRANSLATE/g
        if (re.test(translatedText)) {
          TextStrings[textId] = translatedText
        } else {
          TextStrings[textId] = `${translationHelpTemplate} ${translatedText}`
        }

        TextStrings = JSON.stringify(TextStrings, undefined, 2)
        fs.unlinkSync(path)
        return fs.writeFileSync(path, TextStrings, {encoding: 'utf8'})
      })
    }

    return void 0
  }

  let languageCode = process.argv[2]

  if (!languageCode) { console.log('use: npm run poli -- <languageCode>') }

  if (languageCode) {
    translateLanguageFile(languageCode)
  }
}
templateDir.forEach((filePath) => {
  RunPolina(filePath)
})
