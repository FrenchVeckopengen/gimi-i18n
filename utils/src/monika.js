
const { exec } = require('child_process')
var fs = require('fs')

var languageCodesHolder = ['sv', 'en']
var cleanUpAnnaStore = []
let templateDir = ['./text_strings/client', './text_strings/server', './text_strings/templates', './text_strings/gimi-web']

var textStringsTypes = ['server', 'templates', 'client', 'gimi-web', 'share-image-generator']
var annaTranslationTag = 'ANNA'
var textStrings = {}
var changesMade = false

var runPoli = (key: string):* => {
  return exec(`npm run poli ${key}`)
}

var runAnna = (key: string) :* => {
  return exec(`npm run anna ${key}`)
}

var commitChanges = ():* => {
  console.warn('Commit ')
  require ('child_process').exec('git --rebase', function(err, stdout) {
    console.warn('git pulled ')
    exec('git add --all && git commit -m "translation Bot"')
  })
}

let RunMonika = (filePath):* => {

  let getPath = (file) => `${filePath}/${file}`

  if (file.indexOf('.json') === -1) { return void 0 }

  var path = getPath(file)
  var TextStrings = fs.readFileSync(path, {encoding: 'utf8'})

  TextStrings = JSON.parse(TextStrings)


  textStringsTypes.forEach(textStringsType => {
    textStrings[textStringsType] = {}
  })
  textStringsTypes.forEach(textStringsType => {

    languageCodesHolder.forEach(lang => {
      try {
        textStrings[textStringsType][lang] = require(`../../text_strings/${textStringsType}/${lang}`)
      } catch (e) {
        console.warn(`Cant parse ${textStringsType}/${lang} ${e.message}`)
      }
    })
  })

  textStringsTypes.forEach(textStringsType => {

    languageCodesHolder.forEach(languageCode => {
      var lang = textStrings[textStringsType][languageCode]
      var keys = Object.keys(lang)

      keys.forEach(key => {
        if (languageCode) {
          if (lang[key].includes(annaTranslationTag)) {

            TextStrings[key]lang[key].replace('ANNA', '')
            if (!stringRemoveAnna) {
              console.warn(`Cant find textid: ${textId} in file: ${path}`)
              return void 0
            }
            TextStrings = JSON.stringify(TextStrings, undefined, 2)
            fs.unlinkSync(path)
            return fs.writeFileSync(path, TextStrings, {encoding: 'utf8'})
            
            setTimeout(function () {
              if(languageCode ==='en') {
                console.warn('running poli')
                runPoli(key)
                changesMade = true
              }

              if (languageCode ==='sv') {
                console.warn('running anna')
                runAnna(key)
                changesMade = true
              }
            },300)
          }
        }
      })


    })
  })
  if (changesMade) {

    setTimeout(function () {

      console.warn('Commiting')
      commitChanges()
    }, 900);
  }
}
templateDir.forEach((filePath) => {
  RunMonika(filePath)
})
