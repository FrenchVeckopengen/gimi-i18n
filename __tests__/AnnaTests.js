import {toHash, fromHash, translationHelpTemplate} from '../utils/src/AnnaHelper'

jest.unmock('../utils/src/AnnaHelper')

describe('AnnaHelper', () => {
  it('it should be able to hash string', () => {
    expect(toHash('%1$d har utnyttjas %2$d och har <boldGreen>%3$s</boldGreen> kvar för att få belöningen på %4$s.')).toMatchSnapshot()
  })

  it('it should be able to call fromHash', () => {
    expect(fromHash('89438949034034 har utnyttjas 89438949034035 och har <boldGreen>89438949034032</boldGreen> kvar för att få belöningen på 89438949034033.')).toMatchSnapshot()
  })

  it('it should not have changed translation helper template ', () => {
    expect(translationHelpTemplate).toEqual('*** PLZ_TRANSLATE from en.json ***')
  })

  xit('it should translate from en instead of sv', () => {})
})
