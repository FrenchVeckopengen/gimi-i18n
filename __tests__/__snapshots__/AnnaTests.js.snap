// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AnnaHelper it should be able to call fromHash 1`] = `undefined`;

exports[`AnnaHelper it should be able to hash string 1`] = `undefined`;
